src/sentry/static/sentry/app/views/eventsV2/table/cellAction.tsx

```diff
@@ -0,0 +1,323 @@
+import space from 'app/styles/space';
+import theme from 'app/utils/theme';
+import {tokenizeSearch, stringifyQueryObject} from 'app/utils/tokenizeSearch';
+import {OrganizationSummary} from 'app/types';
+
+import {TableColumn, TableDataRow} from './types';
+
+enum Actions {
+  ADD,
+  EXCLUDE,
+}
+
+type Props = {
+  eventView: EventView;
+  organization: OrganizationSummary;
+  column: TableColumn<keyof TableDataRow>;
+  dataRow: TableDataRow;
+  children: React.ReactNode;
+};
+
+type State = {
+  isHovering: boolean;
+  isOpen: boolean;
+};
+
+export default class CellAction extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    let portal = document.getElementById('cell-action-portal');
+    if (!portal) {
+      portal = document.createElement('div');
+      portal.setAttribute('id', 'cell-action-portal');
+      document.body.appendChild(portal);
+    }
+    this.portalEl = portal;
+    this.menuEl = null;
+  }
+
+  state = {
+    isHovering: false,
+    isOpen: false,
+  };
+
+  componentDidUpdate(_props: Props, prevState: State) {
+    if (this.state.isOpen && prevState.isOpen === false) {
+      document.addEventListener('click', this.handleClickOutside, true);
+    }
+    if (this.state.isOpen === false && prevState.isOpen) {
+      document.removeEventListener('click', this.handleClickOutside, true);
+    }
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('click', this.handleClickOutside, true);
+  }
+
+  private portalEl: Element;
+  private menuEl: Element | null;
+
+  handleClickOutside = (event: MouseEvent) => {
+    if (!this.menuEl) {
+      return;
+    }
+    if (!(event.target instanceof Element)) {
+      return;
+    }
+    if (this.menuEl.contains(event.target)) {
+      return;
+    }
+    this.setState({isOpen: false, isHovering: false});
+  };
+
+  handleMouseEnter = () => {
+    this.setState({isHovering: true});
+  };
+
+  handleMouseLeave = () => {
+    this.setState(state => {
+      // Don't hide the button if the menu is open.
+      if (state.isOpen) {
+        return state;
+      }
+      return {...state, isHovering: false};
+    });
+  };
+
+  handleUpdateSearch = (action: Actions, value: React.ReactText) => {
+    const {eventView, column, organization} = this.props;
+    const query = tokenizeSearch(eventView.query);
+    switch (action) {
+      case Actions.ADD:
+        // Remove exclusion if it exists.
+        delete query[`!${column.name}`];
+        query[column.name] = [`${value}`];
+        break;
+      case Actions.EXCLUDE:
+        // Remove positive if it exists.
+        delete query[`${column.name}`];
+        // Negations should stack up.
+        const negation = `!${column.name}`;
+        if (!query.hasOwnProperty(negation)) {
+          query[negation] = [];
+        }
+        query[negation].push(`${value}`);
+        break;
+      default:
+        throw new Error(`Unknown action type. ${action}`);
+    }
```

Comment:
Should this be extracted to a function so it can be unit tested easier?

Reply:
If the logic were more complex I would. I thought having the functional/integration test was better though as we aren't coupling tests to the details of the component.